### ISSUE TITLE
getScreenResolution is zoom dependent for IE and FF

### ISSUE THREAD

**Issue by mp31415** [Valve/fingerprintjs2#98]
In Chrome getScreenResolution (screen.height, screen.width, screen.availHeight, screen.availWidth) is zoom independent, however in IE and FF these values change upon zoom in/zoom out. (Not sure about Safari/Opera). It makes fingerprint less stable I think. The extra code I use to deal with the issue is the following:

```javascript
getScreenResolution: function(keys) {
	this.getScreenResolutionOrAvail(keys, screen.width, screen.height);
	this.getScreenResolutionOrAvail(keys, screen.availWidth, screen.availHeight);
	return keys;
},
getScreenResolutionOrAvail: function(keys, saw, sah) {
  var res;
  if(saw && sah) {
	  var zoom = 1;
	  if (this.isIE()) {
		  zoom = screen.deviceXDPI / screen.systemXDPI;
		  zoom = this.round5(zoom*100)/100;
	  }
	  else if (this.isFF()) {
		  zoom = window.devicePixelRatio;
	  }
	  if (zoom != 1) {
		sah = this.round10(sah * zoom);
		saw = this.round10(saw * zoom);
	  }
	  res = (sah > saw) ? [sah, saw] : [saw, sah];
  }
  if(typeof res !== "undefined") { // headless browsers
	keys.push(res);
  }
  return keys;
},
round5: function (x)
{
	return Math.round(x/5)*5;
},
round10: function (x)
{
	return Math.round(x/10)*10;
},
```

**Comment by Valve** [comment]
Thanks for the code, last time I checked into this, I couldn't make FF return resolution in a stable way when zooming in/out. I'll look into it.

**Comment by Valve** [comment]
@mp31415 
Thanks for your code, it's a very good start. It returns incorrect resolution for Retina display (tested on FF + MBP 2013 retina), but it's fixable. I have yet to test it on IE browser, but I think it should work there as well. :+1:

**Comment by Valve** [comment]
I've modified your code slightly (mostly cleanup/readability) and added it to the dev branch. The screen resolution now is independent on zoom. However, the canvas FP is different for FF on different zooms. I have opened a separate issue for this.

**Comment by LightCrystal** [comment]
The same problem occurs with Microsoft Edge as well.
the isIE method do not works for Edge, I use this:

var isEdge = function () {
            if (navigator.userAgent.match(/edge/i)) { // Edge
                return true;}
            return false;
};

Testing with IE here:
http://valve.github.io/fingerprintjs2/

I still have the same problem, doing the zoom, the Fingerprint change.
During my internal tests, the resolution is not stable and vary slightly. This rounding seems to not working well at least on my pc.
Windows 10, IE11,

No ZOOM:      2500x1400
150% ZOOM: 2560x1440
400% ZOOM: 2540x1430

**Comment by Valve** [comment]
@LightCrystal thanks for letting me know. I don't have a Windows device with Microsoft Edge on it, only with IE 11 (where zooming is working fine). Please submit a separate issue so that other people with Edge could see it and potentially fix.

**Comment by mp31415** [comment]
@Valve regarding code referencing. Absolutely. We all owe you for this great piece of software. So thanks a lot. 
Regarding Microsoft Edge. I do have Win10/Edge and tested the code (the one I submitted above). It worked fine for me, but my screen resolution is 1920x1200. Cannot test with 2500x1400. This resolution thing becomes messy pretty quick :(.

**Comment by Valve** [comment]
OK, so the issue is big resolutions, where rounding errors start to interfere. Maybe we should add additional scaling factor based on resolution? I don't have a device to test it unfortunately though

**Comment by Valve** [comment]
I've just upgraded my old laptop to Windows 10 and tested on MS Edge. The FP is changing depending on the zoom. The laptop has the resolution of 1600x1200, so I will take a look into it.

**Comment by Valve** [comment]
Moving this issue to v3, because it's much harder to fix reliably, than I anticipated.

**Comment by Valve** [comment]
3 years later, and it's still an issue, which I don't know how to solve reliably. 
This is just a status update. 
We can automatically ignore screen resolution when zoom level is != 1, 
What do you think @jonashaag ?

**Comment by altayevrim** [comment]
Yeah. I think that should be done. I just tested on FF and this still exists.

**Comment by Valve** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/Visual_Viewport_API
This may be how we can solve this in FF.

For IE - we'll be dropping the IE support in V3.

**Comment by Finesse** [comment]
@Valve The issue is still relevant for v3. I think we should solve it.

**Comment by Valve** [comment]
@Finesse I agree, but I don't know how. Also we need to confirm if it's still an issue in Edge based on Chromium.

**Comment by vito2005** [comment]
Hi! Since there's been no activity here for a while, I've picked this up in #1207.

The short version: Firefox reports `screen.width`/`screen.height` in zoom-dependent CSS pixels and rounds them to integers, so multiplying back by `devicePixelRatio` leaves up to ±dpr/2 px of noise (that's why the 2015 approach was unreliable). But Firefox evaluates media queries against the exact fractional size, so the PR recovers it with a binary search over `min-device-width`/`min-device-height` (inspired by [detect-zoom](https://github.com/tombigel/detect-zoom), which uses the same trick for the device pixel ratio) and gets the exact physical resolution regardless of the zoom. When `screen.*` and the media queries disagree (e.g. a spoofing extension) the dimension is reported as unknown to keep the component stable. Details and measurements are in the PR.
